File: Python/MDOF_FRF/MDOF_FRF_Visc_main.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
from MDOF_FRF_Visc_slow import MDOF_FRF_Visc_slow
from MDOF_FRF_Visc import MDOF_FRF_Visc
from MDOF_Eig_Visc import MDOF_Eig_Visc
from N_DOF_sys import N_DOF_sys
from mpl_toolkits.mplot3d import Axes3D
from plot_FRF_3d import plot_FRF_3d
from plot_FRF_Nyq import plot_FRF_Nyq
from plot_FRF_r_i import plot_FRF_r_i
from plot_FRF_mag_phase import plot_FRF_mag_phase

if __name__ == '__main__':
    plt.close('all')
    matplotlib.rcParams['text.usetex'] = True
    
    f_final=40
    n_f_points=8000
    
    N=3
    m_vec=10*np.ones(N)
    c_vec=8*np.ones(N+1)
    k_vec=np.concatenate((161000/2*np.ones(N),[0]), axis=0)
    
    n_row=np.array([0,0,0])
    m_row=np.array([2,1,0])
    
    isproportional=False
    M_mat,C_mat,K_mat=N_DOF_sys(m_vec,c_vec,k_vec)
    #C_mat=0;isproportional=True
    #C_mat=.0001*M_mat+.0001*K_mat;isproportional=True
    
    f_col=np.linspace(0,f_final,n_f_points).reshape(n_f_points,1)
    w_col=2*np.pi*f_col

    #Slow FRF calculation
    H_cols0=MDOF_FRF_Visc_slow(M_mat,C_mat,K_mat,w_col,n_row,m_row)

    #Fast FRF calculation
    EigVectors_Normalized, EigValues_vec=MDOF_Eig_Visc(M_mat, C_mat, K_mat,isproportional)
    H_cols=MDOF_FRF_Visc(EigValues_vec, EigVectors_Normalized, w_col, n_row, m_row)

    print np.amax(np.absolute(H_cols-H_cols0))

    f_3d=plt.figure()
    f_3d.gca(projection='3d')
    
    f_Nyq=plt.figure()
    f_r_i=plt.figure()
    f_mag_phase=plt.figure()
    ax_mag=None
    ax_phase=None
    for ii in range(m_row.size):
        plt.figure(f_3d.number)
        plot_FRF_3d(f_col,H_cols[:,ii],lineLabel=r'$H_{%u,%u}$' %(n_row[ii]+1,m_row[ii]+1),DispRealImag=True)[0]

        plt.figure(f_Nyq.number)
        plot_FRF_Nyq(H_cols[:,ii],lineLabel=r'$H_{%u,%u}$' %(n_row[ii]+1,m_row[ii]+1))[0]
        
        plt.figure(f_r_i.number)
        ax_r,ax_i=plot_FRF_r_i(f_col,H_cols[:,ii],lineLabel=r'$H_{%u,%u}$' %(n_row[ii]+1,m_row[ii]+1))[0:2]
        
        plt.figure(f_mag_phase.number)
        ax_mag,ax_phase=plot_FRF_mag_phase(f_col,H_cols[:,ii],ax_mag,ax_phase,islin_y=False,lineLabel=r'$H_{%u,%u}$' %(n_row[ii]+1,m_row[ii]+1))[0:2]
    
    f_3d.gca().legend(loc='upper left')
    f_3d.savefig('H_3D.pdf')
    f_Nyq.gca().legend()
    f_Nyq.savefig('H_Nyq.pdf', bbox_inches='tight')
    ax_r.legend()
    f_r_i.savefig('H_r_i.pdf', bbox_inches='tight')
    ax_mag.legend()
    f_mag_phase.savefig('H_mag_phase.pdf', bbox_inches='tight')
    
    plt.show()

File: Matlab/MDOF_FRF/MDOF_FRF_Visc_main.m
clearvars
clc
close all

f_final=40;
n_f_points=10000;
islin=0;

N=3;
m=10*ones(1,N);
c=8*ones(1,N+1);
k=[161000/2*ones(1,N),0];

ii_row=[1,1,1];
jj_row=[3,2,1];

[M,C,K]=N_DOF_sys(m,c,k);isProportional=false;
% C=zeros(size(M));isProportional=true;
% C=.0001*M+.0001*K;isProportional=true;

f_col=linspace(0,f_final,n_f_points).';
w_col=2*pi*f_col;

%% Slow FRF
tic
H_cols0=MDOF_FRF_Visc_slow(M,C,K,w_col,ii_row,jj_row);
toc

tic
H_cols1=MDOF_FRF_Visc_Slow_Fortran(M,C,K,w_col,ii_row,jj_row);clear MDOF_FRF_Visc_Slow_Fortran;
toc
max(max(abs(H_cols1-H_cols0)))

tic
H_cols2=MDOF_FRF_Visc_Slow_cpp(M,C,K,w_col,ii_row,jj_row);clear MDOF_FRF_Visc_Slow_cpp;
toc
max(max(abs(H_cols2-H_cols0)))

%% Fast FRF
tic
[EigVectors_Normalized0, EigValues_vec0]=MDOF_Eig_Visc(M, C, K,isProportional);
H_cols3=MDOF_FRF_Visc(EigValues_vec0, EigVectors_Normalized0, w_col, ii_row, jj_row);
toc
max(max(abs(H_cols3-H_cols0)))
H_cols=H_cols3;

tic
[EigVectors_Normalized1, EigValues_vec1]=MDOF_Eig_Visc_Fortran(M, C, K,isProportional);clear MDOF_Eig_Visc_Fortran;
H_cols4=MDOF_FRF_Visc_Fortran(EigValues_vec1, EigVectors_Normalized1, w_col, ii_row, jj_row);clear MDOF_FRF_Visc_Fortran;
toc
max(max(abs(EigVectors_Normalized1-EigVectors_Normalized0)))
max(abs(EigValues_vec1-EigValues_vec0))
max(max(abs(H_cols4-H_cols3)))
H_cols=H_cols4;

tic
[EigVectors_Normalized2, EigValues_vec2]=MDOF_Eig_Visc_cpp(M, C, K,isProportional);clear MDOF_Eig_Visc_cpp;
H_cols5=MDOF_FRF_Visc_cpp(EigValues_vec2, EigVectors_Normalized2, w_col, ii_row, jj_row);clear MDOF_FRF_Visc_cpp;
toc
max(max(abs(EigVectors_Normalized2-EigVectors_Normalized0)))
max(abs(EigValues_vec2-EigValues_vec0))
max(max(abs(H_cols5-H_cols3)))
H_cols=H_cols5;

tic
[EigVectors_Normalized3, EigValues_vec3]=MDOF_Eig_Visc_cpp_Armadillo(M, C, K,isProportional);clear MDOF_Eig_Visc_cpp_Armadillo;
H_cols6=MDOF_FRF_Visc_cpp_Armadillo(EigValues_vec3, EigVectors_Normalized3, w_col, ii_row, jj_row);clear MDOF_FRF_Visc_cpp_Armadillo;
toc
max(max(abs(EigVectors_Normalized3-EigVectors_Normalized0)))
max(abs(EigValues_vec3-EigValues_vec0))
max(max(abs(H_cols6-H_cols3)))
H_cols=H_cols6;

f_3d=figure;hold all

f_r_i=figure;
ax_r=subplot(2,1,1);hold all
ax_i=subplot(2,1,2);hold all

f_Nyq=figure;hold all

f_mag_phase=figure;
ax_mag=subplot(4,1,[1,2,3]);hold all
ax_phase=subplot(4,1,4);hold all

legend_str=cell(length(jj_row),1);
for ii=1:length(jj_row)
    legend_str(ii)=cellstr("$H_{1,"+jj_row(ii)+'}$');
    
    figure(f_3d)
    plot_FRF_3d(f_col,H_cols(:,ii),'','',1);

    plot_FRF_r_i(f_col,H_cols(:,ii),ax_r,ax_i);

    figure(f_Nyq)
    plot_FRF_Nyq(H_cols(:,ii));

    plot_FRF_mag_phase(f_col,H_cols(:,ii),islin,ax_mag,ax_phase);
end

figure(f_3d)
legend(legend_str,'interpreter','latex','Location','bestOutside')

figure(f_Nyq)
legend(legend_str,'interpreter','latex','Location','bestOutside')

legend(ax_r,legend_str,'interpreter','latex','Location','NorthEast')

legend(ax_mag,legend_str,'interpreter','latex','Location','NorthEast')

export_figure((1:4),'',["H_MDOF_3D","H_MDOF_r_i","H_MDOF_Nyq","H_MDOF_mag_phase"])